### PR TITLE
GH-42036: [C++] Add bool operator support to arrow::Result

### PR DESCRIPTION
**Pull request by Kelvinyu1117**
### Rationale for this change
Addressing the comment: from [this PR](https://github.com/apache/arrow/pull/4501#discussion_r293686808)

### What changes are included in this PR?
Add bool operator support to arrow::Result.

### Are these changes tested?
Via CI.

### Are there any user-facing changes?
No

* GitHub Issue: #42036

### CODE REVIEW

File: cpp/src/arrow/result.h [hunk 1]
```diff
@@ -288,7 +288,7 @@ class [[nodiscard]] Result : public util::EqualityComparable<Result<T>> {
   constexpr bool ok() const { return status_.ok(); }
 
   /// \brief Equivalent to ok().
-  // operator bool() const { return ok(); }
+  constexpr explicit operator bool() const { return ok(); }
```
Comment (mapleFU) [inline cpp/src/arrow/result.h:291]: Why a explicit has is added here?

Comment (Kelvinyu1117) [reply cpp/src/arrow/result.h:291]: `explicit` can disable unintended implicit conversions.

Comment (mapleFU) [reply cpp/src/arrow/result.h:291]: This patch LGTM, but considering https://github.com/apache/arrow/pull/4501#discussion_r293686808 . I'll let other decide whether we applying this

**Review by felipecrv — COMMENTED** [review]
Please don't. That's the worst thing that also exists in `std::optional<>` making the use of `std::optional<bool>` a nightmare.

And I can count at least 48 occurrences of `Result<bool>` in the `arrow/cpp/src` directory.

File: cpp/src/arrow/result.h (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/result.h:291]: The explicit way of converting to `bool` is simply calling `ok()`.

`if (result.ok()) { }`, that's all.

**Comment by Kelvinyu1117** [comment]
> Please don't. That's the worst thing that also exists in `std::optional<>` making the use of `std::optional<bool>` a nightmare.
> 
> And I can count at least 48 occurrences of `Result<bool>` in the `arrow/cpp/src` directory.

I see your point, it definitely can be a source of potential bugs for `std::optional<bool>`. Let's see how other people thinks of the pros and cons of this feature.

**Comment by pitrou** [comment]
I agree with @felipecrv . This is a minor convenience that nonetheless risks introducing bugs and confusion.

**Comment by pitrou** [comment]
The (small) consensus here is that this is not desirable, so I'm closing this PR. Still, @Kelvinyu1117 , thank you for taking the time to submit this!

**Pull request closed without merging**